### PR TITLE
Lab Extension Testing

### PR DESCRIPTION
**Pull request by blink1073**
cf #728.

Adds local testing for the lab extension mechanism and tests on Travis.

### CODE REVIEW

**Comment by blink1073** [comment]
This is ready for review.

File: jupyterlab/tests/test_labextensions.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+            install_labextension(self.src, self.name, user=True)
+            self.assert_installed(
+                pjoin(self.name),
+                user=True
+            )
+    
+    def test_create_labextensions_system(self):
+        with TemporaryDirectory() as td:
+            self.system_labext = pjoin(td, u'labextensions')
+            with patch.object(labextensions, 'SYSTEM_JUPYTER_PATH', [td]):
```
Comment (jasongrout) [inline jupyterlab/tests/test_labextensions.py]: You're patching this out in the setup.

File: jupyterlab/tests/test_labextensions.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+                install_labextension(self.src, self.name, user=False)
+                self.assert_installed(
+                    pjoin(self.name),
```
Comment (jasongrout) [inline jupyterlab/tests/test_labextensions.py]: `pjoin(self.name)` is equivalent to `self.name`. This happens other places too.

File: jupyterlab/tests/test_labextensions.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        if user:
+            labext = pjoin(self.data_dir, u'labextensions')
+        else:
+            labext = self.system_labext
+        self.assert_not_dir_exists(
+            pjoin(labext, relative_path)
+        )
+    
+    def test_create_labextensions_user(self):
+        with TemporaryDirectory() as td:
```
Comment (jasongrout) [inline jupyterlab/tests/test_labextensions.py]: Not sure why you need a TemporaryDirectory() call in this function.

File: jupyterlab/tests/test_labextensions.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        enabled = cm.get('jupyterlab_config').get('LabApp', {}).get('labextensions', {}).get(self.name, False)
+        assert enabled
+    
+    def test_labextension_disable(self):
+        self.test_labextension_enable()
+        disable_labextension(self.name)
+        
+        config_dir = os.path.join(_get_config_dir(user=True), 'labconfig')
+        cm = BaseJSONConfigManager(config_dir=config_dir)
+        enabled = cm.get('jupytlab_config').get('LabApp', {}).get('labextensions', {}).get(self.name, False)
```
Comment (jasongrout) [inline jupyterlab/tests/test_labextensions.py]: `jupytlab` -> `jupyterlab`?

**Comment by jasongrout** [comment]
Thanks! This is huge.

File: jupyterlab/tests/build_extension.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+var buildExtension = require('jupyterlab-extension-builder').buildExtension;
+var path = require('path');
+
+
+buildExtension({
+  name: 'mockextension',
+  entryPath: './mockextension/index.js',
+  config: {
+    output: {
+      path: path.join(process.cwd(), 'mockextension', 'build'),
```
Comment (blink1073) [inline jupyterlab/tests/build_extension.js]: Yep, cf https://github.com/jupyter/jupyterlab-extension-builder/issues/8

File: jupyterlab/tests/test_labextensions.py (hunk 1 above)
Comment (blink1073) [inline jupyterlab/tests/test_labextensions.py]: Not this particular directory, I tried removing it and it claimed it did not exist.

File: jupyterlab/tests/build_extension.js (hunk 5 above)
Comment (blink1073) [inline jupyterlab/tests/build_extension.js]: All review comments addressed as comments or in 7d89515

Comment (jasongrout) [inline jupyterlab/tests/build_extension.js]: Fixed in https://github.com/jupyter/jupyterlab-extension-builder/pull/9

File: jupyterlab/tests/test_labextensions.py (hunk 1 above)
Comment (jasongrout) [inline jupyterlab/tests/test_labextensions.py]: This particular directory is patched in the setup: https://github.com/jupyter/jupyterlab/pull/872/files#diff-d0969d946b6c5e183f26f7cc288e6d04R123

Comment (blink1073) [inline jupyterlab/tests/test_labextensions.py]: I have no idea what you mean here.

**Comment by jasongrout** [comment]
I read through it again. Looks good to me; thanks!

**Pull request merged**